packages/fluentui/react-northstar/src/components/Tree/hooks/flattenTree.ts

```diff
@@ -0,0 +1,144 @@
+import { ObjectShorthandCollection, ShorthandRenderFunction, TreeTitleProps } from '../../../index';
+import { TreeItemProps } from '../TreeItem';
+
+export type BaseFlatTreeItem = {
+  /**
+   * Also in TreeItemProps.
+   * The index of the item among its siblings. Count starts at 1.
+   */
+  index: number;
+
+  /**
+   * Also in TreeItemProps.
+   * Level of the tree/subtree that contains this item.
+   */
+  level: number;
+
+  /**
+   * Also in TreeItemProps.
+   * true if the tree item is expanded (indicating tree item has subtree).
+   * when tree item has no subtree, expanded is undefined
+   */
+  expanded?: boolean;
+
+  /**
+   * Also in TreeItemProps.
+   * parent id of the tree item. For the top level tree items, parent is undefined
+   */
+  parent?: string;
+
+  /**
+   * Also in TreeItemProps.
+   * Size of the tree/subtree that contains this item.
+   */
+  treeSize: number;
+
+  /** true if the tree item has subtree, indicating childrenIds are not undefined */
+  hasSubtree: boolean;
+
+  /** children ids of the tree item. For leaf tree item, childrenIds is undefined */
+  childrenIds?: string[];
+};
+
+export type FlatTree = {
+  [key: string]: BaseFlatTreeItem & { [key: string]: any };
+};
+
+export function flattenTree(
+  items: ObjectShorthandCollection<TreeItemProps>, // tree's props.items
+): FlatTree {
+  const result = {};
+  const flatten = (baseItems, level, parent?) => {
+    if (!baseItems) return;
+
+    if (parent) {
+      result[parent].childrenIds = [];
+    }
+
+    baseItems.forEach((item, indexAmongSiblings, arr) => {
+      const { id, items, expanded } = item;
+      if (parent) {
+        result[parent].childrenIds.push(id);
+      }
+
+      result[id] = {
+        index: indexAmongSiblings + 1, // Used for aria-posinset and it's 1-based.
+        level,
+        expanded: items?.length ? !!expanded : undefined,
+        parent: parent == null ? undefined : parent,
+        treeSize: arr.length,
+        hasSubtree: !!items?.length,
+      };
+
+      if (items?.length) {
+        flatten(items, level + 1, id);
+      }
+    });
+  };
+
+  flatten(items, 1);
+  return result;
+}
+
+export function getTobeRenderedItemsProps(
+  items: ObjectShorthandCollection<TreeItemProps>, // tree's props.items
+  flatTree: FlatTree,
+  treeRenderItemTitle?: ShorthandRenderFunction<TreeTitleProps>,
+): TreeItemProps[] {
+  if (!items) {
+    return [];
+  }
+
+  return items.reduce((prevProps: TreeItemProps[], currItem) => {
```

Comment:
Can you please explain why we are iterating there through `items`? `items` is non-flat array, I don't see any guarantee that we will iterate through all items

Reply:
I have to use items here, because flatTree does not keep order.

If I have tree item with id that can be parsed to int, for example '1' '2' ... They will always be in front of all other items in flatTree.

https://stackoverflow.com/questions/5525795/does-javascript-guarantee-object-property-order

Reply:
Resolved by returning an array of ordered ids